File: frontend/lib/password.ts
export const PASSWORD_REQUIREMENTS = [
  { key: "length", label: "At least 8 characters", test: (p: string) => p.length >= 8 },
  { key: "upper", label: "One uppercase letter", test: (p: string) => /[A-Z]/.test(p) },
  { key: "lower", label: "One lowercase letter", test: (p: string) => /[a-z]/.test(p) },
  { key: "digit", label: "One number", test: (p: string) => /\d/.test(p) },
  { key: "special", label: "One special character (!@#$...)", test: (p: string) => /[!@#$%^&*(),.?":{}|<>_\-+=[\]\\;/]/.test(p) },
] as const;

export function validatePassword(password: string): string | null {
  for (const req of PASSWORD_REQUIREMENTS) {
    if (!req.test(password)) return req.label;
  }
  return null;
}

export function getPasswordStrength(password: string): number {
  return PASSWORD_REQUIREMENTS.filter((req) => req.test(password)).length;
}

File: backend/app/utils/password.py
import re

PASSWORD_MIN_LENGTH = 8


def validate_password_strength(password: str) -> str | None:
    """Return an error message if password is weak, else None."""
    if len(password) < PASSWORD_MIN_LENGTH:
        return "Password must be at least 8 characters"
    if not re.search(r"[A-Z]", password):
        return "Password must include an uppercase letter"
    if not re.search(r"[a-z]", password):
        return "Password must include a lowercase letter"
    if not re.search(r"\d", password):
        return "Password must include a number"
    if not re.search(r"[!@#$%^&*(),.?\":{}|<>_\-+=[\]\\;/]", password):
        return "Password must include a special character"
    return None
